### PR TITLE
sdk/metric: Add unit tests for Shutdown WithTimeout

### PR DESCRIPTION
**Pull request by pellared**
EDIT after https://github.com/open-telemetry/opentelemetry-go/pull/4377 is merged: 

Add tests which are are almost copies of existing tests.

### CODE REVIEW

**Review by dmathieu — APPROVED** [review]

File: sdk/metric/periodic_reader.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func (o periodicReaderOptionFunc) applyPeriodic(conf periodicReaderConfig) perio
 
 // WithTimeout configures the time a PeriodicReader waits for an export to
 // complete before canceling it. This includes an export which occurs as part
-// of Shutdown.
+// of ForceFlush and Shutdown.
```
Comment (dashpole) [inline sdk/metric/periodic_reader.go:71]: I'm mostly OK with this, as I think most users will understand what we mean... But technically ForceFlush and Export are different methods on an exporter.  "an export which occurs as part of ForceFlush" might be confusing if ForceFlush does more than export (or doesn't export?).

Comment (pellared) [reply sdk/metric/periodic_reader.go:71]: https://github.com/open-telemetry/opentelemetry-go/blob/c30a8f7649b62201a7782cbeb5fc4a890bfd56f6/sdk/metric/periodic_reader.go#L106-L110

+

https://github.com/open-telemetry/opentelemetry-go/blob/c30a8f7649b62201a7782cbeb5fc4a890bfd56f6/sdk/metric/periodic_reader.go#L297-L299

I think we should improve the doc comment of `ForceFlush` (seperate PR please 😉 ).

Comment (pellared) [reply sdk/metric/periodic_reader.go:71]: The comment was already changed in https://github.com/open-telemetry/opentelemetry-go/pull/4377.

File: sdk/metric/periodic_reader_test.go [hunk 2]
```diff
@@ -329,6 +329,52 @@ func TestPeriodicReaderFlushesPending(t *testing.T) {
 		_ = r.Shutdown(context.Background())
 	})
 
+	t.Run("ForceFlush timeout on producer", func(t *testing.T) {
```
Comment (CharlieTLe) [inline sdk/metric/periodic_reader_test.go:332]: These tests seem to copy how Shutdown is tested. Does it make sense to refactor the shared logic/setup into a reusable function?

Comment (pellared) [reply sdk/metric/periodic_reader_test.go:332]: I was considering it but I did not see a lot of benefits in it as it gave an additional layer of indirection which made the test less readable.

I also try to follow the [Rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))

**Review by MrAlias — APPROVED** [review]

File: sdk/metric/periodic_reader_test.go [hunk 3]
```diff
@@ -329,6 +329,52 @@ func TestPeriodicReaderFlushesPending(t *testing.T) {
...
+				case <-time.After(timeout + time.Second):
+					*rm = testResourceMetricsA
+				case <-ctx.Done():
+					// we timed out before we could collect metrics
+					return ctx.Err()
+				}
+				return nil
+			}})
+		r.RegisterProducer(testExternalProducer{})
+		assert.Equal(t, context.DeadlineExceeded, r.ForceFlush(context.Background()), "timeout error not returned")
```
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/metric/periodic_reader_test.go:348]: Nice. I added it to https://github.com/open-telemetry/opentelemetry-go/issues/4399

File: sdk/metric/periodic_reader.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/periodic_reader.go]: This looks to be resolved.

File: sdk/metric/periodic_reader_test.go (hunk 2 above)
Comment (CharlieTLe) [reply sdk/metric/periodic_reader_test.go:332]: Yeah, I'll give it a go. Thanks :)

**Pull request merged**